edgelet/build/windows/util.ps1

```diff
@@ -1,57 +1,216 @@
+        if ($LastExitCode)
+        {
+            Throw "Failed to download rustup with exit code $LastExitCode"
+        }
+
+        Write-Host "Running rustup-init.exe"
+        ./rustup-init.exe -y --default-toolchain stable-x86_64-pc-windows-msvc
+        if ($LastExitCode)
+        {
+            Throw "Failed to install rust with exit code $LastExitCode"
+        }
+    }
+    else
+    {
+        Write-Host "Running rustup.exe"
+        rustup install stable-x86_64-pc-windows-msvc
+        if ($LastExitCode)
+        {
+            Throw "Failed to install rust with exit code $LastExitCode"
+        }
+    }
+    
+    $ErrorActionPreference = 'Stop'
+}
+
+function InstallWinArmPrivateRustCompiler
+{
+    $link = "https://edgebuild.blob.core.windows.net/iotedge-win-arm32v7-tools/rust-windows-arm.zip"
+
+    Write-Host "Downloading $link"
+    $ProgressPreference = 'SilentlyContinue'
+    Invoke-WebRequest $link -out "rust-windows-arm.zip" -UseBasicParsing
+
+    Write-Host "Extracting $link"
+    Expand-Archive -Path "rust-windows-arm.zip" -DestinationPath "rust-windows-arm"
+    $ProgressPreference = 'Stop'
+}
+
+# arm build has to use a few private forks of dependencies instead of the public ones, in order to to this, we have to 
+# 1. append a [patch] section in cargo.toml to use crate forks
+# 2. run cargo update commands to force update cargo.lock to use the forked crates
+# 3 (optional). when building openssl-sys, cl.exe is called to expand a c file, we need to put the hostx64\x64 cl.exe folder to PATH so cl.exe can be found
+#   this is optional because when building iotedge-diagnostics project, openssl is not required
+function PatchRustForArm
+{
+    Param(
+        [Switch] $OpenSSL
+    )
+
+    $vspath = Join-Path -Path ${env:ProgramFiles(x86)} -ChildPath "Microsoft Visual Studio"
+    Write-Host $vspath
+
+    # arm build requires cl.exe from vc tools to expand a c file for openssl-sys, append x64-x64 cl.exe folder to PATH
+    if($OpenSSL)
+    {
+        try{
+            Get-Command cl.exe -ErrorAction Stop
+        }
+        catch{
+            $cls = (Get-ChildItem -Path $vspath -Filter cl.exe -Recurse -ErrorAction Continue -Force | Sort-Object -Property DirectoryName -Descending)
+            $clPath = ""
+            for ($i= 0; $i -lt $cls.length; $i++) {
+
+                $cl = $cls[$i]
+                Write-Host $cl.DirectoryName
+                if($cl.DirectoryName.ToLower().Contains("hostx64\x64"))
+                {
+                    $clPath = $cl.DirectoryName
+                    break
+                }
+            }
+            $env:PATH = $clPath + ";" + $env:PATH
+            Write-Host $env:PATH
+        }
+
+        # test cl.exe command again to make sure we really have it in PATH
+        Write-Host $(Get-Command cl.exe).Path
+    }
+
+    $edgefolder = Get-EdgeletFolder
+    Set-Location -Path $edgefolder

```

Comment:
Don't change the current directory inside a function.

All cargo commands take a `--manifest-path` parameter, like you can see in other scripts. Use that instead.